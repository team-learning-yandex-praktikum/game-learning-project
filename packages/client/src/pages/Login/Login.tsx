import { FC } from 'react'
import { TextFieldProps } from '@components/TextField/types'
import { Routes } from '@routes/constants'
import AuthForm from '@components/AuthForm'

const fieldsConfig: TextFieldProps[] = [
  {
    name: 'login',
    label: 'Логин',
    placeholder: 'ivan_ivanov',
  },
  {
    name: 'password',
    label: 'Пароль',
    placeholder: 'qwerty',
    type: 'password',
  },
]

export const Login: FC = () => (
  <AuthForm
    title={'Вход'}
    fields={fieldsConfig}
    SubmitButtonProps={{
      children: 'Войти',
    }}
    LinkProps={{
      children: 'Регистрация',
      to: Routes.Registration,
    }}
  />
)
