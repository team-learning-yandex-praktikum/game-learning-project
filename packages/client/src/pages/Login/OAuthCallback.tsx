import {
    ERRCODE_PARAM,
    ERRDESC_PARAM,
    ERROR_NOSTATE,
    ERROR_STATE,
    OAuthApi,
    STATE_PARAM,
    oAuthApi,
} from '@api/oAuth'
import { OauthSignInRequest } from '@api/oAuth/types'
import { Routes } from '@routes/constants'
import { useAppDispatch } from '@store/hooks'
import { loginWithYandex } from '@store/user'
import { ErrorResponse } from '@types'
import { checkState } from '@utils/authentication/oauth'
import { hasKey, isEmptyStr } from '@utils/common/checks'
import { getUrlQueryObj } from '@utils/url/helpers'
import { FC, useCallback, useEffect, useState } from 'react'
import { useNavigate } from 'react-router'

const logError = console.error

const OAuthCallback: FC = () => {
    const [error, setError] = useState('')

    const dispatch = useAppDispatch()
    const navigate = useNavigate()

    const login = useCallback(async (code: string) => {
        try {
            await dispatch(loginWithYandex(code)).unwrap()
            navigate(Routes.Home)
        } catch (e) {
            const err = e as ErrorResponse
            const desc = err.response?.data?.reason ?? ''
            setError(desc)
        }
    }, [])

    useEffect(() => {
        try {
            const queryObj = getUrlQueryObj()

            if (hasKey(ERRCODE_PARAM, queryObj)) {
                const errCode = queryObj[ERRCODE_PARAM]
                const errDesc = queryObj[ERRDESC_PARAM]
                const err = `Ошибка авторизации: ${errDesc} (код ошибки - ${errCode})`
                setError(err)
                return
            }

            if (!hasKey(STATE_PARAM, queryObj)) {
                const err = `Ошибка авторизации: ${ERROR_NOSTATE}`
                setError(err)
                return
            }

            const state = queryObj[STATE_PARAM]
            if (!checkState(state)) {
                const err = `Ошибка авторизации: ${ERROR_STATE}`
                setError(err)
                return
            }

            login(queryObj.code)
        } catch (e) {
            logError(e)
        } finally {
            oAuthApi.cleanup()
        }
    }, [])

    return isEmptyStr(error) ? (
        <div>идет процесс OAuth-авторизации ...</div>
    ) : (
        <div>{error}</div>
    )
}

export { OAuthCallback }
