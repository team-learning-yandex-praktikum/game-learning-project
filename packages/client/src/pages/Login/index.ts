export { default } from './Login'
export { OAuthCallback } from './OAuthCallback'
