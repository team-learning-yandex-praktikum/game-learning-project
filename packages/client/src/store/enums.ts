export enum LoadStatus {
    idle = 'idle',
    loading = 'loading',
    complete = 'complete',
    failed = 'failed',
}
