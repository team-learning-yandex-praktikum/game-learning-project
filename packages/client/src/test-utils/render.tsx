import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { store } from '@store'
import { ReactNode } from 'react'

export const withProviders = (component: ReactNode) => {
    const testStore = store

    return (
        <BrowserRouter>
            <Provider store={testStore}>{component}</Provider>
        </BrowserRouter>
    )
}
