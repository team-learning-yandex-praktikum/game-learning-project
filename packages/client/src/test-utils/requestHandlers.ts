import { http, HttpResponse } from 'msw'
import { joinURL } from '@utils/url/joinURL'
import { EXTERNAL_SERVER_PATH, SERVER_URL } from '@api/baseApi/constants'
import { UserDTO } from '@api/auth/types'
import { transformData } from '@utils/data'
import { UpdatingProfileDTO } from '@api/user/types'

export const userInfoStub: UserDTO = {
    id: 111,
    login: 'ivan',
    display_name: null,
    avatar: null,
    first_name: 'Иван',
    second_name: 'Иванов',
    email: 'ivan@test.ru',
    phone: '79999999999',
}

const baseServerUrl = joinURL(SERVER_URL, EXTERNAL_SERVER_PATH)
export const requestHandlers = {
    signup: http.post(joinURL(baseServerUrl, 'auth/signup'), () =>
        HttpResponse.text(String(userInfoStub.id))
    ),
    signin: http.post(joinURL(baseServerUrl, 'auth/signin'), () =>
        HttpResponse.text('OK')
    ),
    me: http.get(joinURL(baseServerUrl, 'auth/user'), () =>
        HttpResponse.json(userInfoStub)
    ),
    logout: http.post(joinURL(baseServerUrl, 'auth/logout'), () =>
        HttpResponse.text('OK')
    ),
    updatePassword: http.put(joinURL(baseServerUrl, 'user/password'), () =>
        HttpResponse.text('OK')
    ),
    updateProfile: http.put(
        joinURL(baseServerUrl, 'user/profile'),
        async ({ request }) => {
            const body = (await request.json()) as UpdatingProfileDTO
            return HttpResponse.json(
                transformData.from.dto({
                    ...userInfoStub,
                    ...body,
                })
            )
        }
    ),
}
