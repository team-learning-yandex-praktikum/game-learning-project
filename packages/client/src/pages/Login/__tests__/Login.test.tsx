import Login from '../Login'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import { withProviders } from '../../../test-utils/render'
import { setupServer } from 'msw/node'
import { store } from '@store'
import { LoadStatus } from '@store/enums'
import {
    requestHandlers,
    userInfoStub,
} from '../../../test-utils/requestHandlers'
import { fieldsConfig } from '@utils'

const fieldLabels = {
    login: fieldsConfig.login.label,
    password: fieldsConfig.password.label,
    submit: 'Войти',
    link: 'Регистрация',
}

const server = setupServer(requestHandlers.signin, requestHandlers.me)

describe('Login', () => {
    beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close())

    const testedComponent = withProviders(<Login />)

    test('correctly rendered', async () => {
        const { asFragment } = render(testedComponent)

        expect(asFragment()).toMatchSnapshot()
    })

    test('submit should be disabled', () => {
        render(testedComponent)

        const submitButton = screen.getByText(fieldLabels.submit)

        expect(submitButton).toBeDisabled()
    })

    test('should be disabled without required field', async () => {
        render(testedComponent)

        const login = screen.getByLabelText(fieldLabels.login)
        const submitButton = screen.getByText(fieldLabels.submit)

        await userEvent.type(login, userInfoStub.login)

        expect(submitButton).toBeDisabled()
    })

    test('submit should be enabled with correct values', async () => {
        render(testedComponent)

        const login = screen.getByLabelText(fieldLabels.login)
        const password = screen.getByLabelText(fieldLabels.password)
        const submitButton = screen.getByText(fieldLabels.submit)

        await userEvent.type(login, 'ivan')
        await userEvent.type(password, 'SomePass1234')

        expect(submitButton).toBeEnabled()
    })

    test('submit should change state', async () => {
        render(testedComponent)

        const login = screen.getByLabelText(fieldLabels.login)
        const password = screen.getByLabelText(fieldLabels.password)
        const submitButton = screen.getByText(fieldLabels.submit)

        await userEvent.type(login, 'ivan')
        await userEvent.type(password, 'SomePass1234')
        await userEvent.click(submitButton)

        const userState = store.getState().user

        expect(userState?.status).toBe(LoadStatus.complete)
        expect(userState?.data.id).toBe(userInfoStub.id)
    })
})
