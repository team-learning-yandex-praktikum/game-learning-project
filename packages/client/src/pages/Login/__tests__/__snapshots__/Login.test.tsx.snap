// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login correctly rendered 1`] = `
<DocumentFragment>
  <button
    class=""
    type="button"
  >
    <span>
      Войти с помощью Яндекс
    </span>
  </button>
  <div>
    <h2
      class=""
    >
      Вход
    </h2>
    <form
      id="r:id"
    >
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="login"
          >
            Логин
          </label>
          <input
            id="login"
            name="login"
            placeholder="ivan_ivanov"
          />
        </div>
        <div />
      </div>
      <div
        class=""
        data-error="false"
      >
        <div
          class=""
        >
          <label
            for="password"
          >
            Пароль
          </label>
          <input
            id="password"
            name="password"
            placeholder="qwerty"
            type="password"
          />
        </div>
        <div />
      </div>
    </form>
    <div>
      <button
        class=""
        disabled=""
        form="r:id"
        type="submit"
      >
        Войти
      </button>
      <a
        href="/registration"
      >
        Регистрация
      </a>
    </div>
  </div>
</DocumentFragment>
`;
